EventFilter/HGCalRawToDigi/interface/HGCalUnpacker.h

```diff
@@ -0,0 +1,142 @@
+  };
+  enum ECONDHeaderMask {
+    kHeaderMask = 0b111111111,
+    kPayloadLengthMask = 0b111111111,
+    kPassThroughMask = 0b1,
+    kHTMask = 0b11,
+    kEBOMask = 0b11,
+    kMatchMask = 0b1,
+    kTruncatedMask = 0b1,
+  };
+  enum ERXHeaderShift {
+    kFormatShift = 25,
+    kCommonmode0Shift = 15,
+    kCommonmode1Shift = 5,
+  };
+  enum ERXHeaderMask {
+    kFormatMask = 0b1,
+    kCommonmode0Mask = 0b1111111111,
+    kCommonmode1Mask = 0b1111111111,
+  };
+
+  explicit HGCalUnpacker(HGCalUnpackerConfig config);
+
+  /// parse input in S-Link format
+  /// \param[in] inputArray input as 32-bits words vector.
+  /// \param[in] enabledERXMapping map from S-Link indices to enabled eRx in this ECON-D
+  /// \param[in] logicalMapping logical mapping from HGCalElectronicsId to class D as ID
+  void parseSLink(const std::vector<uint32_t>& inputArray,
+                  const std::function<uint16_t(uint16_t sLink, uint8_t captureBlock, uint8_t econd)>& enabledERXMapping,
+                  const std::function<D(HGCalElectronicsId elecID)>& logicalMapping);
+  /// parse input in capture block format
+  /// \param[in] inputArray input as 32-bits words vector.
+  /// \param[in] enabledERXMapping map from capture block indices to enabled eRx in this ECON-D
+  /// \param[in] logicalMapping logical mapping from HGCalElectronicsId to class D as ID
+  void parseCaptureBlock(
+      const std::vector<uint32_t>& inputArray,
+      const std::function<uint16_t(uint16_t sLink, uint8_t captureBlock, uint8_t econd)>& enabledERXMapping,
+      const std::function<D(HGCalElectronicsId elecID)>& logicalMapping);
+  /// parse input in ECON-D format
+  /// \param[in] inputArray input as 32-bits words vector.
+  /// \param[in] enabledERXMapping map from ECON-D indices to enabled eRx in this ECON-D
+  /// \param[in] logicalMapping logical mapping from HGCalElectronicsId to class D as ID
+  void parseECOND(const std::vector<uint32_t>& inputArray,
+                  const std::function<uint16_t(uint16_t sLink, uint8_t captureBlock, uint8_t econd)>& enabledERXMapping,
+                  const std::function<D(HGCalElectronicsId elecID)>& logicalMapping);
+
+  /// \return vector of HGCROCChannelDataFrame<D>(ID, value)
+  const std::vector<HGCROCChannelDataFrame<D> >& channelData() const { return channelData_; }
+  /// \return vector of 32-bit index, the length is the same as channelData(), link from channel data to the first common mode on ROC (+0,+1,+2,+3 for all four common modes)
+  const std::vector<uint32_t>& commonModeIndex() const { return commonModeIndex_; }
+  /// \return vector of 16-bit common mode data, lowest 10 bits is the ADC of the common mode, padding to 4 for half ROC turned on
+  const std::vector<uint16_t>& commonModeData() const { return commonModeData_; }
+  const std::vector<uint32_t>& badECOND() const { return badECOND_; }
+
+private:
+  const uint32_t erxBodyLeftShift_[16] = {2, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
```

Comment:
Are these numbers (from `L114` to `L132`) something that could change along time? If this is the case, wouldn't be better to add them in a `cfg` file?

Reply:
@clacaputo Does this address your concern about the `const`-ness of these numbers?

Reply:
@forthommel , it does. @yulunmiao thanks for the clarification.